rust/main/chains/hyperlane-ethereum/src/tx.rs

```diff
@@ -219,20 +220,25 @@ where
 pub(crate) async fn call_with_lag<M, T>(
     call: ethers::contract::builders::ContractCall<M, T>,
     provider: &M,
-    maybe_lag: Option<NonZeroU64>,
+    lag: &ReorgPeriod,
 ) -> ChainResult<ethers::contract::builders::ContractCall<M, T>>
 where
     M: Middleware + 'static,
     T: Detokenize,
 {
-    if let Some(lag) = maybe_lag {
-        let fixed_block_number: BlockNumber = provider
-            .get_block_number()
-            .await
-            .map_err(ChainCommunicationError::from_other)?
-            .saturating_sub(lag.get().into())
-            .into();
-        Ok(call.block(fixed_block_number))
+    if !lag.is_none() {
+        let reorg_period = EthereumReorgPeriod::try_from(lag)?;
+        let block = match reorg_period {
+            EthereumReorgPeriod::Blocks(lag) => provider
+                .get_block_number()
+                .await
+                .map_err(ChainCommunicationError::from_other)?
+                .saturating_sub(lag.into())
+                .into(),
+            EthereumReorgPeriod::Tag(tag) => tag,
```

Comment:
in that case I think the problem is in `get_finalized_block_number` - it shouldn't make the network call if `tag` is `BlockId::Number`, but I still think we can reuse `get_finalized_block_number` here

Reply:
sorry should've been more clear @apastushenka, I'd like you to add this optimization in `get_finalized_block_number` to avoid the extra network call while reusing its logic